### PR TITLE
fix!: improve unimarkup-inline API and implementation

### PR DESCRIPTION
**Pull request by nfejzic**
# Goal of PR

This PR aims to improve API and (hopefully) performance of `unimarkup-inline` crate. 

# Inline structs instead of shared `InlineContent`

Previously we used `InlineContent` enum with two content types (`Nested` and `Plain`) for inlines. This way, we could control which inlines have nested content (contain other inlines) and which are simple strings. However, inlines do not necessarily share everything, so customizing content makes more sense. 

This PR replaces `InlineContent` shared among many `Inline` variants with static dispatch approach, similar to `Block` in `unimarkup-core`. This means that each `Inline` variant gets a corresponding struct (i.e. `Inline::Bold(Bold { ... })`). Using this approach we get better type safety (i.e. `render_bold` now gets `Bold` instead of `NestedContent` etc.).

There are more or less two groups of inline types: plain and nested inlines. Plain inlines have `String` as their content, whereas nested inlines contain other inlines. They are now stored in separate modules. Since inlines in one group share a lot of implementation, they're implemented with macros.  

Also, some other parts got a little bit cleaner and easier to reason about (for example merging of inlines, inlines construction and so on). 

## No alloation during lexing of inlines

Another change is borrowing instead of allocating content during lexing of inlines. Previously, plain tokens stored `String` as their content. These tokens are later consumed by the parser and inlines are constructed, where we allocate a `String` again. This means we had a redundant allocation that was not really necessary. 

In this PR, `Token` now borrows a `str` from `Symbol`, and all the machinery surrounding tokenization do that as well. This means that nothing should be allocated during tokenization, and only allocation happens during construction of inlines. Consequently, this should mean that all element parsers that use inlines parser should have no allocations, and all allocations happen during parsing of inlines. 

In order to make this possible, content of single `Token` must come from **consecutive** `Symbol`s. So, whenever we encounter escape sequences (e.g. `\*\*`) we will have multiple tokens (i.e. one token **up to** the escape symbol `\` and one token **from** the escape symbol). Of course, this means that snapshots for lexer now look a little different. For parser it makes little difference (start - end positions might be different, this is open for discussion though).

### CODE REVIEW

File: inline/src/lexer/mod.rs [hunk 1]
```diff
@@ -558,7 +569,9 @@ impl Iterator for Tokens {
             self.iter.next()?
         };
 
-        match unr_token.pop() {
+        let part = unr_token.pop();
+
+        match part {
```
Comment (mhatzl) [inline inline/src/lexer/mod.rs:574]: why this change?
maybe leftover from debugging?

File: inline/src/parser/mod.rs [hunk 2]
```diff
@@ -28,84 +26,117 @@ impl Deref for ParserStack {
...
+    fn parse_plain(&mut self, start_token: Token, enclosed: bool) -> Inline {
+        let kind = start_token.kind;
+        let (tkn_str, mut span) = start_token.parts();
 
-        if kind != TokenKind::Plain && !token.opens() {
-            let (content, span) = token.into_inner();
-            let content = InlineContent::Plain(PlainContent::new(content, span));
-            return Inline::as_plain_or_eol(content, kind);
-        }
+        // dbg!(&start_token);
```
Comment (mhatzl) [inline inline/src/parser/mod.rs:64]: dead code?

**Review by mhatzl — COMMENTED** [review]
very nice work. Especially the macro for plain and nested structs.
I found some commented code, and I suggest to rename

- `Newline` to `ExplicitNewline` since there seems to be a mixup between `Newline` and `EndOfLine`
- `Whitespace` to `ExplicitWhitespace`
- `EndOfLine` to `Newline`
- introduce `Whitespace` we still need to know that this grapheme is a whitespace to compress contiguous whitespaces to one

**Comment by nfejzic** [comment]
> introduce Whitespace we still need to know that this grapheme is a whitespace to compress contiguous whitespaces to one

What would be the best way to do this? 

I think we could add a function to check if some token is opening of an inline with preserved whitespace (i.e. Verbatim) and in that case we store them "as is". For all other (plain) inlines (that don't preserve whitespace) we can compress them (i.e. Skip all but one whitespace).

**Comment by mhatzl** [comment]
> > introduce Whitespace we still need to know that this grapheme is a whitespace to compress contiguous whitespaces to one
> 
> What would be the best way to do this?
> 
> I think we could add a function to check if some token is opening of an inline with preserved whitespace (i.e. Verbatim) and in that case we store them "as is". For all other (plain) inlines (that don't preserve whitespace) we can compress them (i.e. Skip all but one whitespace).

Not sure.
I'd say storing each whitespace grapheme as its own `Whitespace` token when lexing, and then combine them to one space during parsing if not in preserved-whitespace context or explicit escape.
Maybe we do not need a function as you describe, depending how parsing of verbatim etc is done. If verbatim simply flattens all symbols, then all whitespaces will be preserved anyways.

**Comment by nfejzic** [comment]
> If verbatim simply flattens all symbols, then all whitespaces will be preserved anyways.

Yes, but in non-whitespace-preserving context we should compress them. That's why some kind of check is needed (e.g. `token.preserves_whitespace()` or something similar).

**Comment by mhatzl** [comment]
Ah ok. I thought we always compress, unless content is flattened or explicitly escaped.
But it might be more clear by using a function to check how whitespace should be handled.

File: inline/src/inlines/mod.rs [hunk 3]
```diff
@@ -1,231 +1,285 @@
...
+            }),
+            _ => Self::Plain(Plain {
+                content: content.into(),
+                span,
+            }),
+        }
+    }
+
+    /// Creates a new [`Inline::Plain`] from the given content with the given [`Span`].
+    pub fn plain(content: String, _kind: TokenKind, span: Span) -> Self {
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/inlines/mod.rs:112]: I think you're right. One edge case that occurred to me is, currently it is not possible to have an explicit newline in a `plain` inline. I don't know if that's relevant for LSP implementation. For example, verbatim inline would not contain any explicit newlines, because the content is simply a String...

File: inline/src/inlines/types/plain.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    Plain,
+    /// A verbatim inline with input preserved.
+    Verbatim,
+    /// Content inside of pair of parentheses.
+    Parentheses,
+    /// Newline literal.
+    Newline,
+    /// Any whitespace literal except newline.
+    Whitespace,
+    /// Inline explicitly marking end of line.
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/inlines/types/plain.rs:57]: That would make lexing a little simpler IMO. Right now, consecutive plain symbols are consumed into single `Token` with `TokenKind::Plain`. This consumes whitespace as well. But this was done at lexing stage previously because of `String` allocations, larger plain tokens would reduce number of allocations at parsing stage. But there are no allocations at lexing stage now, so we can do this differently. 

My suggestion would be, if lexer does not recognize symbol as a keyword, substitution or an escape sequence, we can simply wrap the symbol into plain token and pass it to parser. Parser can then decide whether whitespace should be preserved or merged. 🤔

File: inline/src/lexer/mod.rs [hunk 5]
```diff
@@ -402,9 +402,10 @@ impl TokenIterator<'_> {
     ///
     /// [`Token`]: self::token::Token
     /// [`TokenKind::Plain`]: self::token::TokenKind::Plain
-    fn lex_plain(&mut self) -> Option<Token> {
-        let start_pos: position::Position = self.get_symbol(self.index)?.start;
-        let mut content = String::with_capacity(self.symbols.len());
+    fn lex_plain(&mut self) -> Option<Token<'token>> {
+        // let mut content = String::with_capacity(self.symbols.len());
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/lexer/mod.rs:406]: Nice catch!

File: inline/src/lexer/mod.rs (hunk 1 above)
Comment (nfejzic) [reply inline/src/lexer/mod.rs:573]: Correct!

**Comment by nfejzic** [comment]
# Update

This PR is now ready for review, `logid` changes are merged from main, `unimarkup-inline` implementation is (in my opinion) improved, testing is slightly changed. In other words, a lot of changes in a big PR. 

## Inline improvements

There are multiple changes in inlines, mostly described in the PR description. Some additional are, we can now compress multiple consecutive whitespace tokens into a single whitespace. Previously, plain token included textual content together with whitespaces, so it was not possible to do. Now, whitespace tokens are created as well, meaning plain text is split into multiple tokens (with whitespace as separators). No allocations in lexing stage change helps here, because we would have a lot of string allocations otherwise. 

## Testing changes

In implementation, we sometimes have a `unreachable!` macro call, indicating that this state should not be possible. If such statement is reached, then we immediately know that there's a bug in implementation, and code panics (also tests). 
For testing, we create dynamic snapshot tests. Meaning, we load tests from files, and adding new tests means add new files, or new test cases in existing files.

But, when a test panics in a way other than snapshot test failing, we do not see which test case caused that. Only the wrapping function (e.g. `test_lexer_snapshots`), and that makes it hard to hunt the bug down. 

This can be solved in two ways: 

1. Create a `proc_macro` that finds all test files recursively in a given directory, creates test case functions and annotates them with `#[test]` attribute. I did not want to do this, because `proc_macro`s are not that easy to understand, and also it feels like *magic*. Better to avoid them if possible...
2. Use a [custom test harness](https://www.infinyon.com/blog/2021/04/rust-custom-test-harness/). This works great, especially when combined with [cargo-nextest](https://nexte.st/).

I chose to go with the custom harness option. Combined with `cargo-nextest` it looks great! It does look a little *non-standard* when run with `cargo t` and some tests panic, but I still think it's great because at the end we get list of failed test-cases, and also panic message. However, there are some limitations, like output not being captured this way (i.e. `println!`, `dbg!` and so on will print out during test runs). Therefore this should really be constrained to use cases where it makes sense (snapshot tests). 

Here's an example how a test run with above setup and `cargo nextest run` looks like (`parser` part is to filter only parser tests):

image

If I forgot to explain something, please write a comment and I'll try to clarify 😄

**Comment by mhatzl** [comment]
Starting the review, I got errors in VS Code (rust analyzer), but `cargo t` ran without issues with rust version `1.68.0`.
After updating to version `1.71.1`, rust analyzer showed no errors anymore.

Maybe we should increase the minimum rust version required for Unimarkup.

File: inline/tests/spec/markup/verbatim.yml [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+tests:
+  - name: simple-verbatim
+    description: |
+      Simple verbatim content.
+
+    input: |
+      `Verbatim`
+
+    html: |
+      <code>Verbatim</code>
```
[the remark this answers is not in the document]
Comment (mhatzl) [reply inline/tests/spec/markup/verbatim.yml:16]: Using `<pre>` might be better to keep whitespaces as is.

**Review by mhatzl — CHANGES_REQUESTED** [review]
Very nice improvements overall 👍 

Most comments from my side are about handling of whitespaces.
The mix between explicit/escaped and unescaped whitespace is still a bit mixed in the code.
As discussed, it might be good to look how we can improve this in general, and for the LSP highlighting.

About your testing approach:\
It seems really comfortable to define new test cases now, because you don't even have to specify the files to test anymore.
I think this would be nice to have for all snap- and spec-tests, but I leave it up to you if you want to change that in another PR.

Thanks :)

File: inline/src/parser/mod.rs [hunk 7]
```diff
@@ -2,110 +2,181 @@ use std::{collections::VecDeque, ops::Deref};
...
+            let not_enclosed_and_interrupted = !cxt.enclosed && next_token.kind != kind;
+
+            if enclosed_and_closes {
+                span.end = next_token.span.end;
+                break;
+            } else if not_enclosed_and_interrupted {
+                if next_token.is_plain_whitespace() {
+                    // consume whitespace
+                    let (next_content, next_span) = next_token.parts();
+                    content.push_str(next_content);
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/parser/mod.rs:92]: We do check already with `is_whitespace` in scanner (see `symbol` impl in `unimarkup_commons` crate). I agree about pushing plain space 👍🏻

File: inline/src/parser/mod.rs [hunk 8]
```diff
@@ -2,110 +2,181 @@ use std::{collections::VecDeque, ops::Deref};
...
+                // opening token not yet closed, next token does not start a new inline -> parse as
+                // plain text
 
                 end = next_token.span.end;
-                content.append(InlineContent::from(next_token));
+                let (inner, span) = next_token.parts();
+                let inline = Inline::Plain(Plain {
+                    content: String::from(inner),
+                    span,
+                });
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/parser/mod.rs:142]: Oh that's a nice catch 🦅 👀. Generalisation at this point would probably be the correct choice. Instead of arbitrarily creating a plain inline, we should call `parse_plain` similarly how we do it for `parse_inline` few lines above. Maybe even completely remove check for `next_token.opens()` and call `parse_inline` regardless if token closes or is some content that should simply be taken as is.

File: inline/src/parser/mod.rs [hunk 9]
```diff
@@ -2,110 +2,181 @@ use std::{collections::VecDeque, ops::Deref};
...
+        // opening/closing of tokens is resolved at lexing stage
+        // at this point we can simply parse
+        let kind = token.kind;
+
+        let mut inline = match (kind, token.opens()) {
+            (TokenKind::Verbatim, _) => self.parse_plain(
+                token,
+                PlainContext {
+                    enclosed: true,
+                    merge_whitespace: false,
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/parser/mod.rs:161]: I am not entirely sure if we should use this for some other inlines apart from verbatim, but otherwise you're right. Right now, that naming would make more sense.

[the remark this answers is not in the document]
Comment (mhatzl) [reply inline/src/parser/mod.rs:161]: I am also unsure. Maybe adding doc-comments is sufficient enough.

File: commons/src/scanner/symbol.rs [hunk 10]
```diff
@@ -176,7 +176,10 @@ impl Symbol<'_> {
     /// It is assumed (and checked in debug release) that the symbols are in contiguous order.
     ///
     /// Returns `None` if the referenced input is not same in all symbols.
-    pub fn flatten_iter<'s>(mut iter: impl Iterator<Item = &'s Symbol<'s>>) -> Option<&'s str> {
+    pub fn flatten_iter<'o, 's>(mut iter: impl Iterator<Item = &'o Symbol<'s>>) -> Option<&'s str>
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply commons/src/scanner/symbol.rs:179]: Good observation. I will try to rename the lifetimes so that naming is more consistent. Also, I will document them in doc comments, as that should help with understanding (and works nice in IDE hover pop-up documentation etc).

File: inline/src/inlines/substitute.rs
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/inlines/substitute.rs:1]: Issues would be nice! Thanks!

File: inline/src/inlines/types/nested.rs [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+                &self.content
+            }
+
+            /// Returns mutable reference to inner content.
+            pub fn inner_mut(&mut self) -> &mut VecDeque<Inline> {
+                &mut self.content
+            }
+
+            /// Merges consecutive inlines of same kind contained in this inline.
+            pub fn try_merge(&mut self) {
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/inlines/types/nested.rs:41]: That's a very valid critique. I will take a look into this and think a little more about it. Maybe rename to `merge(&mut self)` would suffice, as this cannot fail (simply does nothing if merging not possible) and we don't need to know if it succeeded or not.

File: inline/src/inlines/mod.rs [hunk 12]
```diff
@@ -1,231 +1,281 @@
...
     /// [`InlineContent`]: self::content::InlineContent
-    pub fn new(content: InlineContent<PlainContent, NestedContent>, kind: TokenKind) -> Self {
-        let consume_as_plain = |content| match content {
-            InlineContent::Plain(plain_content) => Self::Plain(plain_content),
-            InlineContent::Nested(nested_content) => Self::Multiple(nested_content),
-        };
+    pub fn plain_or_eol(content: impl Into<String>, span: Span, kind: TokenKind) -> Self {
+        let content = content.into();
+        match kind {
+            TokenKind::Newline => Self::EndOfLine(EndOfLine { content, span }),
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/inlines/mod.rs:101]: I hear your question, and I give you a question: Why not? 🤣 

Joking aside, this is just a result of confusing naming scheme I used. 
Would following naming be alright: 

* `Newline` - regular newline in content (not escaped newline)
* `Whitespace` - regular whitespace in content, like space or tab (not escaped whitespace)
* `EscapedNewline` - newline that is escaped (i.e. `\\n`)
* `EscapedWhitespace` - whitespace that is escaped (e.g. `\\t`)

This way we would now what means what.

File: inline/src/lexer/token.rs [hunk 13]
```diff
@@ -150,10 +144,21 @@ impl Token {
...
+                | TokenKind::Whitespace
+                | TokenKind::ExplicitWhitespace
         )
     }
 
+    /// Checks whether this [`Token`] represents plain whitespace content (i.e. space or tab).
+    ///
+    /// [`Token`]: self::Token
+    pub fn is_plain_whitespace(&self) -> bool {
+        matches!(self.kind, TokenKind::Whitespace)
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/lexer/token.rs:159]: The idea of this function is to check whether a given token is a whitespace that can be consumed by a plain inline. This is important when we want to merge multiple whitespace tokens, or if a plain inline starts with that kind of token (implementation detail). Do you have recommendations for a better name? 

I will see if there's a way to solve this in a more generic way 🤔. Maybe the way to go is two functions: 

1. `consumable_by(&self, inline: &Inline) -> bool` which tests whether the given token (`&self`) can be consumed as content of the given inline
2. `can_merge_with(&self, other: &Token) -> bool` which tests whether the given token (`&self`) can merge with the `other` token. This would also enable merging of different kinds of tokens, make it more generic etc, but also complicate a little bit even though we don't have a need for that right now. 

What do you think?

File: inline/tests/spec/markup/verbatim.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/tests/spec/markup/verbatim.yml:16]: I would leave that for a separate PR.

File: inline/src/parser/mod.rs [hunk 14]
```diff
@@ -2,110 +2,181 @@ use std::{collections::VecDeque, ops::Deref};
...
 
-        if kind == TokenKind::Plain {
-            content.append(InlineContent::from(token));
+        while let Some(next_token) = self.next_token() {
+            let enclosed_and_closes = cxt.enclosed && next_token.closes(Some(&start_token));
+            let not_enclosed_and_interrupted = !cxt.enclosed && next_token.kind != kind;
+
+            if enclosed_and_closes {
+                span.end = next_token.span.end;
+                break;
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/parser/mod.rs:87]: Because the `next_token` closes the given `Inline` and is not needed except for `Span` information (as seen in code). Basically, we've found the end, and we do not that end except for it to signal that it is the end (of some `Inline`).

File: inline/src/parser/mod.rs [hunk 15]
```diff
@@ -2,110 +2,181 @@ use std::{collections::VecDeque, ops::Deref};
...
+        Inline::nested_with_span(content, start_token.kind, span)
+    }
+
+    fn parse_inline(&mut self, token: Token) -> Inline {
+        // opening/closing of tokens is resolved at lexing stage
+        // at this point we can simply parse
+        let kind = token.kind;
+
+        let mut inline = match (kind, token.opens()) {
+            (TokenKind::Verbatim, _) => self.parse_plain(
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/parser/mod.rs:157]: `Verbatim` is now also *kind-of* plain inline, because it just consumes any content found inside as a string. The context passed to `parse_plain` indicates what to do with consecutive whitespaces, which should handle the merging of whitespaces. 

We discussed internally how to represent multi-line strings so that we can also keep information about line-wise span information. That could potentially make it possible to handle multi-line verbatim inline in a better way (GitHub treats newlines in verbatim as a whitespace).

File: inline/src/parser/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/parser/mod.rs:161]: I will add doc-comments here and leave it as is for now.

File: inline/tests/lexer/snapshot.rs [hunk 16]
```diff
@@ -46,6 +46,14 @@ impl AsSnapshot for Snapshot<Token> {
             "\n" => "\u{23CE}",
             other => other,
         };
+
+        if span.len_utf8().unwrap_or(1).saturating_sub(inner.len()) == 1 {
+            // escaped token's occupy two characters in text (the backslash and symbol). In such
+            // cases, span is longer than the actual content by a single character.
+            // Push content to the right "\*" will be rendered as " *"
+            content.push(' ');
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/tests/lexer/snapshot.rs:54]: I agree, will use *the blank* symbol: ␢ (U+2422).

File: inline/src/inlines/types/nested.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (mhatzl) [reply inline/src/inlines/types/nested.rs:41]: `merge(&mut self)` sounds good

File: inline/src/inlines/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (mhatzl) [reply inline/src/inlines/mod.rs:101]: Yes that sounds good 👍

File: inline/src/lexer/token.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (mhatzl) [reply inline/src/lexer/token.rs:159]: uh I like that 👍 
maybe add `consumable_by_plain(&self) -> bool`, which should nearly be a drop in replacement of `is_plain_whitespace`, right?

File: inline/src/parser/mod.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (mhatzl) [reply inline/src/parser/mod.rs:157]: GitHub follows the commonmark spec (https://spec.commonmark.org/0.30/#inlines), where `<code>` tags are used for inline verbatim, but `<pre><code>` for verbatim blocks. To still preserve non line-breaking whitespace in inline verbatim, they recommend adding CSS `code{white-space: pre-wrap;}`. 

I am still unsure if we should do the same, or use `<pre><code>` in both cases, but convert a line-break to space for inline verbatim. For now, I will create an issue in the spec repo.

You may skip this in your PR.

File: inline/src/inlines/types/plain.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/inlines/types/plain.rs:57]: I still didn't name them as I should've. Here's how will that look like now: 

```
Newline -> just regular newline
Whitespace -> any regular whitespace
EscapedNewline -> Newline that has been explicitly escaped (e.g. `\\n`)
EscapedWhitespace -> Whitespace that has been explicitly escaped (e.g. `\\t`)
```

This will now be same for `TokenKind` and `Inline`.

File: inline/src/inlines/substitute.rs
[the remark this answers is not in the document]
Comment (mhatzl) [reply inline/src/inlines/substitute.rs:1]: created issue #101 for this

File: inline/src/lexer/token.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/lexer/token.rs:159]: Yes that is correct. I will introduce function you suggest, and later on if the need arises to check in a more generic way we can change it.

File: inline/src/lexer/token.rs [hunk 17]
```diff
@@ -394,28 +403,32 @@ pub enum TokenKind {
     /// Double colon for substitution (`::`).
     Substitution,
 
-    /// Escaped newline token (`\\n`).
+    /// End of line - regular newline token ('\n').
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/lexer/token.rs:406]: I now used `Escaped` for escaped tokens and inlines, and without prefix for the non-escaped ones.

File: inline/src/parser/mod.rs [hunk 18]
```diff
@@ -2,110 +2,181 @@ use std::{collections::VecDeque, ops::Deref};
...
 
     fn deref(&self) -> &Self::Target {
         &self.data
     }
 }
 
+struct PlainContext {
+    enclosed: bool,
+    merge_whitespace: bool,
+}
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/parser/mod.rs:26]: Added doc comments.

File: inline/src/parser/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/parser/mod.rs:92]: This is also a potential one where edge cases can be solved with a function on Token. Something like: `fn content_for(&self, inline: &Inline) -> &str`. The changes like `consumable_by_plain` function etc. make this a little tedious to handle right now, but since it's the single edge case for now I would keep a little tedious but easy to understand for now.

File: inline/src/parser/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply inline/src/parser/mod.rs:142]: This is simplified a lot now, thank you for the observation and remarks over video call. The function now only checks if closing token is found, otherwise calls `parse_inline` making this significantly simpler.

File: inline/tests/lexer/snapshot.rs [hunk 19]
```diff
@@ -43,9 +43,17 @@ impl AsSnapshot for Snapshot<Token> {
...
         let inner = match self.as_str() {
-            "\n" => "\u{23CE}",
+            "\n" => Self::NEWLINE_SYBMOL,
             other => other,
         };
+
+        if span.len_utf8().unwrap_or(1).saturating_sub(inner.len()) == 1 {
+            // escaped token's occupy two characters in text (the backslash and symbol). In such
+            // cases, span is longer than the actual content by a single character.
+            // Push content to the right "\*" will be rendered as " *"
```
Comment (mhatzl) [inline inline/tests/lexer/snapshot.rs:53]:
```suggestion
            // e.g. content like "\*" will be rendered as "␢*" in snapshots to indicate the backslash escape.
```

Just to clarify that we only render it like that for testing.

File: inline/tests/spec/snapshots/parser/bold/escaped-bold.snap [hunk 20]
```diff
@@ -1,9 +1,10 @@
 ---
 source: inline/tests/parser/mod.rs
-info: "Test 'escaped-bold' from: 'bold.yml'"
+info: "Test 'escaped-bold' from 'bold.yml'"
 ---
 Plain @ (1:1)->(1:17) (
-    **not bold**
+    ␢␢**not bold**␢␢
```
Comment (mhatzl) [inline inline/tests/spec/snapshots/parser/bold/escaped-bold.snap:6]:
```suggestion
    ␢*␢*not bold*␢*␢
```

Would it be difficult to render it like this?
Just to show where a backslash was actually set.
It would also be consistent with the lexer snapshot.

File: inline/tests/spec/snapshots/parser/verbatim/escaped-verbatim.snap [hunk 21]
```diff
@@ -0,0 +1,15 @@
+---
+source: inline/tests/parser/mod.rs
+info: "Test 'escaped-verbatim' from 'verbatim.yml'"
+---
+Plain @ (1:1)->(1:17) (
+    ␢`not verbatim`␢
```
Comment (mhatzl) [inline inline/tests/spec/snapshots/parser/verbatim/escaped-verbatim.snap:6]: see comment for escaped bold snapshot

File: inline/tests/spec/snapshots/parser/verbatim/simple-verbatim.snap [hunk 22]
```diff
@@ -0,0 +1,15 @@
+---
+source: inline/tests/parser/mod.rs
+info: "Test 'simple-verbatim' from 'verbatim.yml'"
+---
+Verbatim @ (1:1)->(1:11) (
+    ␢Verbatim␢
```
Comment (mhatzl) [inline inline/tests/spec/snapshots/parser/verbatim/simple-verbatim.snap:6]: why are there ␢?
If it is intentional, I think we should search for another symbol and use ␢ only to represent a backslash.
Otherwise it might be confusing to read.

File: inline/tests/spec/snapshots/parser/verbatim/verbatim-in-middle.snap [hunk 23]
```diff
@@ -0,0 +1,23 @@
+---
+source: inline/tests/parser/mod.rs
+info: "Test 'verbatim-in-middle' from 'verbatim.yml'"
+---
+Plain @ (1:1)->(1:10) (
+    The next
+    ^^^^^^^^^
+)
+Verbatim @ (1:10)->(1:16) (
+    ␢word␢
```
Comment (mhatzl) [inline inline/tests/spec/snapshots/parser/verbatim/verbatim-in-middle.snap:10]: see comment in previous file

File: inline/tests/spec/snapshots/parser/verbatim/verbatim-not-verbatim.snap [hunk 24]
```diff
@@ -0,0 +1,19 @@
+---
+source: inline/tests/parser/mod.rs
+info: "Test 'verbatim-not-verbatim' from 'verbatim.yml'"
+---
+Verbatim @ (1:1)->(1:11) (
+    ␢Verbatim␢
```
Comment (mhatzl) [inline inline/tests/spec/snapshots/parser/verbatim/verbatim-not-verbatim.snap:6]: see comment in `simple_verbatim` file

**Review by mhatzl — COMMENTED** [review]
Very nice work 👍 

I found some minor mistakes in doc-comments, and I am not sure about using ␢ in verbatim snapshots.
Otherwise, looks great :)

File: inline/tests/spec/snapshots/parser/bold/escaped-bold.snap (hunk 20 above)
Comment (nfejzic) [reply inline/tests/spec/snapshots/parser/bold/escaped-bold.snap:6]: Unfortunately yes. Depending on the test case maybe even completely impossible. 
This is because at this point we only have a single `String` of content, and no information where the `Span` in input and output length mismatch happened.

Comment (mhatzl) [reply inline/tests/spec/snapshots/parser/bold/escaped-bold.snap:6]: no problem :)
Lets leave it as is then

File: inline/tests/spec/snapshots/parser/verbatim/simple-verbatim.snap (hunk 22 above)
Comment (nfejzic) [reply inline/tests/spec/snapshots/parser/verbatim/simple-verbatim.snap:6]: This is indeed intentional. The goal is to use `␢` as a way to show the mismatch between `Span` some element (e.g. `Inline`) occupies in input text, and the length of the actual content of that element. This way we can easily see that the calculated `Span` is indeed correct, even though the output is *"shorter"* than what the `Span` suggests.

Comment (mhatzl) [reply inline/tests/spec/snapshots/parser/verbatim/simple-verbatim.snap:6]: fine with me 👍 
(just a bit unfortunate that `␢` != escaped using **b**ackslash `\` 😄)

**Review by mhatzl — APPROVED** [review]
Looks good 👍 
Thanks again :)
